cmd/kubeadm/app/util/runtime/runtime.go

```diff
@@ -38,64 +41,89 @@ var defaultKnownCRISockets = []string{
 
 // ContainerRuntime is an interface for working with container runtimes
 type ContainerRuntime interface {
-	Socket() string
+	Connect() error
+	SetImpl(impl)
 	IsRunning() error
 	ListKubeContainers() ([]string, error)
 	RemoveContainers(containers []string) error
 	PullImage(image string) error
 	PullImagesInParallel(images []string, ifNotPresent bool) error
-	ImageExists(image string) (bool, error)
+	ImageExists(image string) bool
 	SandboxImage() (string, error)
 }
 
 // CRIRuntime is a struct that interfaces with the CRI
 type CRIRuntime struct {
-	exec       utilsexec.Interface
-	criSocket  string
-	crictlPath string
+	impl           impl
+	logger         *klog.Logger
+	criSocket      string
+	runtimeService criapi.RuntimeService
+	imageService   criapi.ImageManagerService
 }
 
+// defaultTimeout is the default timeout inherited by crictl
+const defaultTimeout = 2 * time.Second
+
 // NewContainerRuntime sets up and returns a ContainerRuntime struct
-func NewContainerRuntime(execer utilsexec.Interface, criSocket string) (ContainerRuntime, error) {
-	const toolName = "crictl"
-	crictlPath, err := execer.LookPath(toolName)
-	if err != nil {
-		return nil, errors.Wrapf(err, "%s is required by the container runtime", toolName)
+func NewContainerRuntime(criSocket string) ContainerRuntime {
+	logger := klog.Background()
```

Comment:
what are the implications of this?
would the root --v of kubeadm control the verbosity?
can it be spammy at lower levels and is there info the average user will care about?

with crictl we didn't have addition verbosity, and wondering if we should just 'nil' this.

Reply:
Yeah, let's use `nil` for now (same for the tracer provider. We can pass a logger if we want to have increased verbosity in any case later on.

Reply:
Makes sense, removed it.